lib/SILOptimizer/Transforms/AllocBoxToStack.cpp

```diff
@@ -274,38 +284,78 @@ static bool partialApplyEscapes(SILValue V, bool examineApply) {
+                                unsigned CurrentRecurDepth) {
   SILFunction *F = ApplySite(O->getUser()).getReferencedFunctionOrNull();
   // If we cannot examine the function body, assume the worst.
   if (!F || F->empty())
     return false;
 
-  // We don't actually use these because we're not recursively
-  // rewriting the partial applies we find.
-  SmallVector<Operand *, 1> PromotedOperands;
+  // Since this function can be called recursively while analyzing the same box,
+  // mark the callee as visited, so that we don't end up in a recursive cycle.
+  auto iter = VisitedCallees.insert(F);
+  if (!iter.second)
+    return false;
+
   auto calleeArg = F->getArgument(ApplySite(O->getUser()).getCalleeArgIndex(*O));
-  return !findUnexpectedBoxUse(calleeArg, /* examinePartialApply = */ false,
-                               /* inAppliedFunction = */ true,
-                               PromotedOperands);
+  auto res = !findUnexpectedBoxUseInDFSOrder(calleeArg,
+                                             /* inAppliedFunction = */ true,
+                                             PromotedOperands, VisitedCallees,
+                                             CurrentRecurDepth + 1);
+  return res;
+}
+
+// Returns true if a callee is eligible to be cloned and rewritten for
+// AllocBoxToStack opt. We don't want to increase code size, so this is
+// restricted only for private local functions currently.
+static bool isOptimizableApplySite(ApplySite Apply) {
+  auto callee = Apply.getReferencedFunctionOrNull();
+  if (!callee) {
+    return false;
+  }
+
+  // Callee should be optimizable.
+  if (!callee->shouldOptimize())
+    return false;
+
+  // External function definitions.
+  if (!callee->isDefinition())
+    return false;
+
+  // Do not optimize always_inlinable functions.
+  if (callee->getInlineStrategy() == Inline_t::AlwaysInline)
+    return false;
+
+  if (callee->getLinkage() != SILLinkage::Private)
+    return false;
+
+  return true;
 }
 
 /// Validate that the uses of a pointer to a box do not eliminate it from
-/// consideration for promotion to a stack element. Optionally examine the body
-/// of partial_apply to see if there is an unexpected use inside.  Return the
-/// instruction with the unexpected use if we find one.
+/// consideration for promotion to a stack element. Return the instruction with
+/// the unexpected use if we find one.
+/// If a box has ApplySite users, we recursively examine the callees to check
+/// for unexpected use of the box argument. If all the callees through which the
+/// box is passed don't have any unexpected uses, `PromotedOperands` will be
+/// populated with the box arguments in DFS order.
 static SILInstruction *
-findUnexpectedBoxUse(SILValue Box, bool examinePartialApply,
-                     bool inAppliedFunction,
-                     SmallVectorImpl<Operand *> &PromotedOperands) {
+findUnexpectedBoxUseInDFSOrder(SILValue Box, bool inAppliedFunction,
```

Comment:
My thought:

```
recursivelyFindBoxOperandsPromotableToAddress
```

The reason I am pushing on this is that the important thing here is computing PromotedOperands. That we return the unexpected use is just a side-effect. Your thoughts?

That being said. The comment is beautiful and is exactly what I hoped for!